dm-agent/src/batcher.rs

```diff
@@ -0,0 +1,415 @@
+
+        Ok(())
+    }
+
+    pub fn add_to_batch(&mut self, message: Message) -> Result<(), DeviceMonitorError> {
+        let collectd_message = CollectdMessage::parse_from(&message)?;
+
+        self.message_grouper.measurement(
+            Some(collectd_message.metric_group_key),
+            collectd_message.metric_key,
+            collectd_message.metric_value,
+        )?;
+
+        Ok(())
+    }
+
+    pub fn end_batch(&mut self) -> Option<MeasurementGrouper> {
+        if self.message_grouper.is_empty() {
+            None
+        } else {
+            Some(std::mem::replace(
+                &mut self.message_grouper,
+                MeasurementGrouper::new(),
+            ))
+        }
+    }
+}
+
+pub struct MessageBatcher {
+    sender: UnboundedSender<MeasurementGrouper>,
+    mqtt_client: Arc<dyn MqttClient>,
+    topic_filter: TopicFilter,
+}
+
+impl MessageBatcher {
+    pub fn new(
+        sender: UnboundedSender<MeasurementGrouper>,
+        mqtt_client: Arc<dyn MqttClient>,
+    ) -> Result<Self, DeviceMonitorError> {
+        let topic_filter = TopicFilter::new(SOURCE_TOPIC)?.qos(QoS::AtMostOnce);
+        Ok(Self {
+            sender,
+            mqtt_client,
+            topic_filter,
+        })
+    }
+
+    pub async fn run(&self) -> Result<(), DeviceMonitorError> {
+        let mut errors = self.mqtt_client.subscribe_errors();
+        tokio::task::spawn(async move {
+            while let Some(error) = errors.next().await {
+                error!("MQTT error: {}", error);
+            }
+        });
+
+        let mut messages = self
+            .mqtt_client
+            .subscribe(self.topic_filter.clone())
+            .await?;
+
+        loop {
+            match self.run_message_batcher(messages.as_mut()).await {
+                Ok(_) => {
+                    //If the message batching loop returns, it means the MQTT connection has closed
+                    error!("MQTT connection closed");
+                    break;
+                }
+                Err(error) => {
+                    error!("Error while building a message batch: {}", error);
+                }
+            }
+        }
+        Ok(())
+    }
+
+    async fn run_message_batcher(
+        &self,
+        messages: &mut dyn MqttMessageStream,
+    ) -> Result<(), DeviceMonitorError> {
+        let batching_window = Duration::from_millis(DEFAULT_STATS_COLLECTION_WINDOW);
+        let mut message_batch = MessageBatch::new();
+        while let Some(message) = messages.next().await {
+            //Start a new message batch
+            message_batch.start_batch(message)?;
+
+            //Add messages to the current batch until the batching window times out and return
+            self.build_message_batch_with_timeout(&mut message_batch, messages, batching_window)
+                .await?;
+
+            //Close the current batch and send it to the batch processor
+            if let Some(batch) = message_batch.end_batch() {
```

Comment:
One will never see an empty batch here, since we added a message line 147.

This is not a big issue, but this highlight that the `message_batch` should be created & populated by the `build_message_batch_with_timeout` method.